Actor groups and group sync helpers

AndrewDryga opened the pull request.

AndrewDryga committed:
Add manually managed actor groups

AndrewDryga committed:
Fix logger warning

AndrewDryga committed:
Add mechanics for recurrent jobs

AndrewDryga committed:
Implement Google Directory API client

AndrewDryga committed:
Add missing config helper

AndrewDryga committed:
Add Google Workspace adapter

AndrewDryga committed:
Dump updated_at on upserts

AndrewDryga committed:
Remove unused function

AndrewDryga committed:
Make credo happy

AndrewDryga committed:
Prevent timing attacks on IdP state

AndrewDryga committed:
Fix Phoenix LiveReload for the domain app

AndrewDryga committed:
Add browser timezone and locale for future use

AndrewDryga committed:
tmp

AndrewDryga committed:
Add created_by fields to all relevant schemas

AndrewDryga committed:
Change auth provider behaviour and add missing domain logic

AndrewDryga committed:
Warn when config changed in dev mode

AndrewDryga committed:
Change status page widget to use lookup table for classes

AndrewDryga committed:
Add CLDR library

AndrewDryga committed:
Mount account ID from path params

AndrewDryga committed:
Fix seeds

AndrewDryga committed:
Cleanup component library for navigation, layout and breadcrumbs

AndrewDryga committed:
Refactor router to remove path duplication and group routes in sections

AndrewDryga committed:
Fix compilation race condition with protocol consolidation

AndrewDryga committed:
Finish UI around Identity Providers

AndrewDryga committed:
Fix some wip artifacts

AndrewDryga commented on elixir/apps/domain/lib/domain/actors.ex:
```diff
@@ -1,8 +1,93 @@
+    change_group(%Group{}, attrs)
+  end
+
+  def upsert_provider_groups(%Auth.Provider{} = provider, attrs_by_provider_identifier) do
```
This function will be finished in a separate PR.

AndrewDryga commented on elixir/apps/domain/test/support/fixtures/actors_fixtures.ex:
```diff
@@ -3,6 +3,66 @@ defmodule Domain.ActorsFixtures do
+    group
+  end
+
+  # def create_provider_group(attrs \\ %{}) do
```
This will be finished in a separate PR.

AndrewDryga committed:
Fix post-rebase issues

AndrewDryga committed:
Fix Chromedriver CI failures

AndrewDryga committed:
Make credo happy

AndrewDryga committed:
Make dialyzer happy

AndrewDryga committed:
Add missing domain tests

AndrewDryga committed:
Remove test token from form component

jamilbk commented on elixir/apps/domain/lib/domain/jobs/executors/global.ex:
```diff
@@ -0,0 +1,164 @@
+    # `random_notify_name` is used to avoid name conflicts in a cluster during deployments and
+    # network splits, it randomly selects one of the duplicate pids for registration,
+    # and sends the message {global_name_conflict, Name} to the other pid so that they stop
+    # tying to claim job queue leadership.
```
```suggestion
    # trying to claim job queue leadership.
```

jamilbk commented on elixir/apps/domain/lib/domain/actors/group/changeset.ex:
```diff
@@ -0,0 +1,58 @@
+  end
+
+  # We do not update the `name` field because we allow to manually override it in the UI
+  # for usability reasons when the provider uses group names that can make people confused
```
👍

bmanifold replied:
Just wanted to make sure I'm following here....Is this saying that when we're syncing groups from a customers IDP it won't be syncing the name of the group if the admin has manually set the name in Firezone?

jamilbk commented on elixir/apps/domain/lib/domain/actors.ex:
```diff
@@ -1,8 +1,93 @@
+    end
+  end
+
+  def create_group(attrs, %Auth.Subject{} = subject) do
```
Doesn't have to be in this PR (apologies if I missed it), but just checking if we are still planning to lock actor memberships to groups of the same provider? I.e. so we avoid weird edge cases with manually-added actors to provider-synced groups.

AndrewDryga replied:
@jamilbk people should be able to edit groups managed manually, but it should not let you manually manage synced stuff.

AndrewDryga replied:
I will add a couple of tests to cement this decision in the code.

jamilbk commented on elixir/apps/domain/lib/domain/auth.ex:
Nicely architected 😎

jamilbk commented on elixir/apps/domain/lib/domain/auth.ex:
```diff
@@ -261,6 +404,7 @@ defmodule Domain.Auth do
     salt = Keyword.fetch!(config, :salt)
     # TODO: we don't want client token to be invalid if you reconnect client from a different ip,
     # for the clients that move between cellular towers
+    # TODO: we want to show all sessions in a UI so persist them to DB
```
👍 Will be required for audit reports in some form, could start with a basic export to CSV function as an MVP

jamilbk commented on elixir/apps/domain/lib/domain/auth/adapters/google_workspace.ex:
```diff
@@ -0,0 +1,75 @@
+      end
+    )
+
+    # TODO: validate received scopes and show an error if there are missing ones
```
claims?

```suggestion
    # TODO: validate received claims and show an error if there are missing ones
```

jamilbk commented on elixir/apps/domain/lib/domain/jobs/executors/global.ex:
```diff
@@ -0,0 +1,164 @@
+
+      UPDATE my_table SET processing_cancelled_at = NOW() WHERE id = ?;
+
+  Even though this does not prove fully fledged job queue, it is a good enough solution
```
```suggestion
  Even though this does not provide a fully fledged job queue, it is a good enough solution
```

jamilbk commented on elixir/apps/domain/lib/domain/jobs/executors/global.ex:
```diff
@@ -0,0 +1,164 @@
+    # tying to claim job queue leadership.
+    with :no <- :global.register_name(name, self(), &:global.random_notify_name/3),
+         pid when is_pid(pid) <- :global.whereis_name(name) do
+      # we monitor the leader process so that we start a race to become a new leader with it's down
```
```suggestion
      # we monitor the leader process so that we start a race to become a new leader when it's down
```

jamilbk commented on elixir/apps/domain/lib/domain/jobs/executors/global.ex:
```diff
@@ -0,0 +1,164 @@
+  end
+
+  # tick is scheduled by using a timeout message instead of `:timer.send_interval/2`,
+  # because we jobs to overlap if they take too long to execute
```
```suggestion
  # because we don't want jobs to overlap if they take too long to execute
```

jamilbk commented on elixir/apps/web/lib/web/components/layouts/app.html.heex:
Thank you for cleaning this up 😅

jamilbk reviewed (approved):
Just some minor typo suggestions and a couple questions, but otherwise this beast LGTM

jamilbk commented on .github/workflows/elixir.yml:
```diff
@@ -305,7 +305,7 @@ jobs:
         options: --cap-add=IPC_LOCK
     steps:
-      - uses: nanasess/setup-chromedriver@v1
+      - uses: nanasess/setup-chromedriver@v2
```
Fixed in #1827 to unblock other PRs

bmanifold commented on elixir/apps/domain/lib/domain/actors/actor/changeset.ex:
```diff
@@ -1,16 +1,15 @@
+    |> put_change(:account_id, account_id)
   end
 
-  def set_actor_type(actor, type) when type in [:account_user, :account_admin_user] do
```
Just curious why these are no longer needed?

AndrewDryga replied:
@bmanifold Nothing changed in practice, before we had a match (the function will raise an error if an invalid value is passed), not we will get a changeset error because it's an enum value. I removed it here because don't want to duplicate enum values across the codebase.

bmanifold replied:
Gotcha, that makes sense.  Thanks for the info!

bmanifold commented on elixir/apps/domain/lib/domain/auth.ex:
```diff
@@ -24,6 +24,62 @@ defmodule Domain.Auth do
+    end
+  end
+
+  def fetch_provider_by_id(id) do
```
I feel like I've noticed a lot of the other models use exclamation marks at the end of methods that aren't checking permissions.  Just wondering if this method could use one as well?

AndrewDryga replied:
Exclamation is a convention that means that function will raise, this function returns `:ok` and `:error` tuples so it should not have one.

bmanifold replied:
Ah, ok, cool.  Very good to know.  I had interpreted the exclamation mark to mean "use with caution".

bmanifold commented on elixir/apps/domain/lib/domain/config/definitions.ex:
```diff
@@ -444,13 +444,13 @@ defmodule Domain.Config.Definitions do
-    default: ~w[email openid_connect token]a
+    {:array, ",", {:parameterized, Ecto.Enum, Ecto.Enum.init(values: ~w[
+      email
+      openid_connect google_workspace
```
Just curious if these two were meant to be on the same line?

AndrewDryga replied:
It doesn't matter much, I keep them together because they are basically the same (google extends oidc).

bmanifold replied:
Ah, ok, sounds good. 👍

bmanifold commented on elixir/apps/web/lib/web/auth.ex:
```diff
@@ -169,10 +207,14 @@ defmodule Web.Auth do
     {:cont, mount_subject(socket, params, session)}
   end
 
+  def on_mount(:mount_account, params, session, socket) do
+    {:cont, mount_account(socket, params, session)}
+  end
+
   def on_mount(:ensure_authenticated, params, session, socket) do
     socket = mount_subject(socket, params, session)
 
-    if socket.assigns.subject do
+    if socket.assigns[:subject] do
```
For my own edification, is this change a best practice or is there an actual difference in the way the "subject" is being accessed here?

AndrewDryga replied:
If the subject is not assigned before the function will just return whatever is in `else` block, but now it will raise. I don't think it should be ever called when the subject is not assigned, so it will help to fight some accidental mistakes.

bmanifold replied:
Ah, ok, interesting.

bmanifold commented on elixir/apps/web/lib/web/components/core_components.ex:
```diff
@@ -221,7 +190,7 @@ defmodule Web.CoreComponents do
       page={5}
       total_pages={100}
-      collection_base_path={~p"/users"}/>
+      collection_base_path={~p"/actors"}/>
```
Are all references to "users" now going to be "actors" from now on?  I ask because my current PR likely will need to be changed if so.

AndrewDryga replied:
This is hard again because of naming, we can keep them as `users` but then our domain language doesn't match our front-end, but if we name them Actors everywhere it can be confusing for our end users 🤷‍♂️.

bmanifold replied:
Yeah, definitely agree this is a tough naming problem.

Will service/machine accounts be listed along side human users?  Or will they have their own URL path?

jamilbk replied:
I was planning to have them in a separate area in the UI, so it probably makes sense to scope the routes then?

bmanifold replied:
Well, the reason I was asking was that I had the thought that if everything is under one page, then having the URL have `actors` in it might not be a big deal, but if things are going to get split out with having a page of only human users then it may be better to have the URL say `users`? 🤷‍♂️

jamilbk replied:
@bmanifold I was thinking the same thing. `users` and `service_accounts` may be the same thing in the backend, but from an Admin's perspective, they're really two different things with different security contexts and workflows.

jamilbk replied:
Hmm, but this makes for more UIs for us to build. I suppose we could keep things simpler for now, and just have one `actors` UI section with controls for segregating `service_account` actors and `user` actors.

On that note, we may also want to update `device` to `client` like @AndrewDryga had mentioned; I agree that `Device` is not a great name for headless clients running via `service_accounts`.

jamilbk replied:
We could discuss more on Slack or a quick call

AndrewDryga committed:
Test that manually created groups can not be managed edited

AndrewDryga committed:
Better error messages for discovery_document_uri validation

AndrewDryga committed:
Fix typos

bmanifold commented on elixir/apps/web/lib/web/components/core_components.ex:
```diff
@@ -713,6 +547,94 @@ defmodule Web.CoreComponents do
+  end
+
+  @doc """
+  Helps to pluralize a word based on a an cardinal number.
```
```suggestion
  Helps to pluralize a word based on a cardinal number.
```

AndrewDryga committed:
Add missing domain tests and functions

bmanifold reviewed (approved):
Nothing blocking.  LGTM.  Just had a few small questions.

AndrewDryga committed:
Add tests for LV modules and new IdP controllers

AndrewDryga committed:
Remove live suffix in module and file names

It's all already in Live folder, no need to duplicate it 3 times (folder, file name, module name).

AndrewDryga committed:
Fix typos

AndrewDryga committed:
Remove even more suffixes

AndrewDryga committed:
Fix typo in a controller

AndrewDryga committed:
Make dialyzer happy
